### PR TITLE
kraken: Add new method "has_code_type" in PtRef Grammar

### PR DESCRIPTION
**Pull request by benoit-bst**
Add new method in PtRef Grammar, `has_code_type` like

```
filter=collection.has_code_type(Type)
```
First part for **JIRA** : https://jira.kisio.org/browse/NAVP-1219

### CODE REVIEW

File: source/ptreferential/ptreferential_utils.cpp [hunk 1]
```diff
@@ -309,6 +309,46 @@ type::Indexes get_indexes_from_code(const type::Type_e type,
...
+        nt::CodeContainer::SupportedTypes,
+        T>::type,
+    Indexes>::type
+get_indexes_from_codes(const std::string& key, const Data& data) {
+    Indexes indexes;
+    auto collection = data.pt_data->collection<T>();
+    for (const auto* obj: collection) {
+        auto codes = data.pt_data->codes.get_codes<T>(obj);
+        if (codes.empty() || codes.find(key) == codes.end()) { continue; }
+            indexes.insert(obj->idx);
```
Comment (xlqian) [inline source/ptreferential/ptreferential_utils.cpp]: weird indentation 
```suggestion
        indexes.insert(obj->idx);
```

File: source/ptreferential/ptreferential_utils.h [hunk 2]
```diff
@@ -65,5 +65,8 @@ type::Indexes get_indexes_from_id(const type::Type_e type,
 type::Indexes get_indexes_from_name(const type::Type_e type,
                                     const std::string& name,
                                     const type::Data& data);
+type::Indexes get_indexes_from_codes(const type::Type_e type,
```
Comment (xlqian) [inline source/ptreferential/ptreferential_utils.h]: you mean `get_indexes_from_codes_type`?

File: source/ptreferential/ptreferential_utils.cpp [hunk 3]
```diff
@@ -309,6 +309,46 @@ type::Indexes get_indexes_from_code(const type::Type_e type,
...
+}
+type::Indexes get_indexes_from_codes(const type::Type_e type,
+                                     const std::string& key,
+                                     const type::Data& data) {
+    switch (type) {
+#define GET_INDEXES(type_name, collection_name) \
+        case Type_e::type_name: return get_indexes_from_codes<type::type_name>(key, data);
+        ITERATE_NAVITIA_PT_TYPES(GET_INDEXES)
+#undef GET_INDEXES
+    default: return Indexes();// no code supported, empty result
```
Comment (xlqian) [inline source/ptreferential/ptreferential_utils.cpp]: Indexes{}, just like you did at L338 :)
```suggestion
    default: return Indexes{};// no code supported, empty result
```

File: source/ptreferential/ptreferential_utils.cpp [hunk 4]
```diff
@@ -309,6 +309,46 @@ type::Indexes get_indexes_from_code(const type::Type_e type,
...
+    typename boost::mpl::contains<
+        nt::CodeContainer::SupportedTypes,
+        T>::type,
+    Indexes>::type
+get_indexes_from_codes(const std::string& key, const Data& data) {
+    Indexes indexes;
+    auto collection = data.pt_data->collection<T>();
+    for (const auto* obj: collection) {
+        auto codes = data.pt_data->codes.get_codes<T>(obj);
+        if (codes.empty() || codes.find(key) == codes.end()) { continue; }
```
Comment (xlqian) [inline source/ptreferential/ptreferential_utils.cpp]:
```suggestion
        if (codes.find(key) == codes.end()) { continue; }
```

File: source/ptreferential/tests/ptref_test.cpp [hunk 5]
```diff
@@ -558,6 +558,46 @@ BOOST_AUTO_TEST_CASE(code_request) {
...
+    res = make_query(nt::Type_e::StopPoint,
+                     R"(stop_point.has_code_type(code_type_2))",
+                     *(b.data));
+    BOOST_CHECK_EQUAL_RANGE(get_uris<nt::StopPoint>(res, *b.data), std::set<std::string>({"stop2", "stop4"}));
+
+    res = make_query(nt::Type_e::StopArea,
+                     R"(stop_area.has_code_type(code_type_1))",
+                     *(b.data));
+    BOOST_CHECK_EQUAL_RANGE(get_uris<nt::StopArea>(res, *b.data), std::set<std::string>({"sa_1"}));
+}
```
Comment (kadhikari) [inline source/ptreferential/tests/ptref_test.cpp:603]: It is preferable to add a test a query on absent code type (for example on code_type_3 ?)

**Review by kadhikari — APPROVED** [review]

File: documentation/rfc/ptref_grammar.md [hunk 6]
```diff
@@ -63,6 +63,7 @@ In the following table, if the invocation starts with `collection`, any collecti
 | query | signification | note |
 |-------|---------------|------|
 |`collection.has_code(type, value)`|all the objects of type `collection` that have the code `{type: "type", value: "value"}` in `codes[]`|only the `codes[]` field is used|
+|`collection.has_code_type(type)`|all the objects of type `collection` that have this kind of code `{type: "type"}` in `codes[]`|only the `codes[]` field is used|
```
Comment (nonifier) [inline documentation/rfc/ptref_grammar.md]:
```suggestion
|`collection.has_code_type(type)`|all the objects of type `collection` that has a code with the specific type `{type: "type"}` in `codes[]`|only the `codes[]` field is used|
```

File: source/ptreferential/ptreferential_utils.cpp [hunk 7]
```diff
@@ -309,6 +309,46 @@ type::Indexes get_indexes_from_code(const type::Type_e type,
...
+}
+template<typename T>
+static
+typename boost::disable_if<
+    typename boost::mpl::contains<
+        nt::CodeContainer::SupportedTypes,
+        T>::type,
+    Indexes>::type
+get_indexes_from_code_type(const std::string&, const Data&) {
+    // there is no codes for unsupporded types, thus the result is empty
```
Comment (nonifier) [inline source/ptreferential/ptreferential_utils.cpp]:
```suggestion
    // there is no code for unsupported types, thus the result is empty
```

**Review by nonifier — APPROVED** [review]

**Comment by benoit-bst** [comment]
Jenkins, test this please !

**Comment by benoit-bst** [comment]
Jenkins, test this please !

**Pull request merged**